File: Regression/hotCppR/driver.cpp
//*********************************************************
// This file is part of Rapsodia released under the LGPL. *
// The full COPYRIGHT notice can be found in the top      *
// level directory of the Rapsodia distribution           *
//*********************************************************
#include <vector>
#include <cmath>
#include <cassert>
#include <iostream>
#include <iomanip>
#include "RAinclude.ipp"
#include "HigherOrderTensor.hpp"

double explicitSinProdTensorEntry(unsigned short n,
				  std::vector<double> x,
				  std::vector<unsigned short> index) { 
  double result=1.0;
  for(unsigned short i=1; i<=n;i++) { 
    switch(index[i-1]%4) { 
    case (0) :
      result*=sin(x[i-1]);
      break; 
    case (1) :  
      result*=cos(x[i-1]);
      break; 
    case (2) :
      result*=(-sin(x[i-1])); 
      break;
    case (3) :
      result*=(-cos(x[i-1]));
      break;
    default : 
      assert(0);
      break;
    }
  }
  return result; 
}

int main() { 
  int rc=0;
  double myEps=1.0E-12;
  unsigned short  n,o;
  n=3;
  o=3;	
  HigherOrderTensor T(n,o); 
  int dirs=T.getDirectionCount();
  int i,j,k;
  // argument values
  std::vector<RAfloatD> x(n);
  std::vector<double> xv(n);
  for(i=1;i<=n;i++) { 
    xv[i-1]=1.0+.1*i;
    x[i-1]=xv[i-1];
  }
  // get the seed matrix
  Matrix<unsigned int> SeedMatrix=T.getSeedMatrix();
  for(i=1;i<=dirs;i++) { 
    for(j=1;j<=n;j++) { 
      x[j-1].set(i,1,SeedMatrix[j-1][i-1]);
    }
  }
  // compute the target function
  extern void head(const std::vector<RAfloatD>& x,
		   RAfloatD& y) ;
  RAfloatD y;
  head(x,y);
  // transfer the taylor coefficients
  Matrix<double> TaylorCoefficients(o,dirs);
  for(i=1;i<=o;i++) { 
    for(j=1;j<=dirs;j++) { 
      TaylorCoefficients[i-1][j-1]=y.get(j,i);
    }
  }
  T.setTaylorCoefficients(TaylorCoefficients);
  // harvest the compressedTensor
  for (k=1; k<=o; k++){ 
    std::cout << "order: " << k << std::endl;
    double entry=0.0;
    std::vector<double> compressedTensor=T.getCompressedTensor(k);
    HigherOrderTensor Helper(n,k); 
    dirs=Helper.getDirectionCount();
    // get the helper seed matrix
    Matrix<unsigned int> HelperSeedMatrix=Helper.getSeedMatrix();
    for(i=1;i<=dirs;i++) {
      std::vector<unsigned short> index(n);
      for(j=1;j<=n;j++) { 
	index[j-1]=HelperSeedMatrix[j-1][i-1];
      }
      entry=explicitSinProdTensorEntry(n,xv,index);
      std::cout << "T";
      for(j=1;j<=n;j++) { 
      std::cout << "[" << std::setw(2) << HelperSeedMatrix[j-1][i-1] << "]";
      }
      std::cout << compressedTensor[i-1] << " ?= " << entry ;
      if (fabs(compressedTensor[i-1]-entry)>myEps) { 
	std::cout << " diff is: " << fabs(compressedTensor[i-1]-entry);
        rc=1;
      } 
      else { 
        std::cout << " ok"; 
      }
      std::cout << std::endl;
    }
  }
  return rc;
}

File: Regression/hotF90R/driver.f90
!#########################################################
! This file is part of Rapsodia released under the LGPL. #
! The full COPYRIGHT notice can be found in the top      #
! level directory of the Rapsodia distribution           #
!#########################################################
double precision function explicitSinProdTensorEntry(n,x,index)
  implicit none
  integer :: n
  double precision :: x(n)
  integer :: index(n)
  integer i
  explicitSinProdTensorEntry=1.0
  do i=1,n
     select case (MOD(index(i),4))
        case (0) 
           explicitSinProdTensorEntry=explicitSinProdTensorEntry*sin(x(i))
        case (1) 
           explicitSinProdTensorEntry=explicitSinProdTensorEntry*cos(x(i))
        case (2) 
           explicitSinProdTensorEntry=explicitSinProdTensorEntry*(-sin(x(i)))
        case (3) 
           explicitSinProdTensorEntry=explicitSinProdTensorEntry*(-cos(x(i)))
        case default
           stop 'no logic for this case'
     end select
  end do
end function explicitSinProdTensorEntry


program driver
  include 'RAinclude.i90'

  use higherOrderTensorUtil

  implicit none 

  external head


  type(RArealD), dimension(:), allocatable :: x
  real(kind=RAdKind), dimension(:), allocatable :: xv
  type(RArealD) :: y
  integer n,d
  type(higherOrderTensor) :: T,Thelper
  integer, dimension(:,:), allocatable :: SeedMatrix,SeedMatrixHelper
  real(kind=RAdKind), dimension(:,:), allocatable :: TaylorCoefficients
  real(kind=RAdKind), dimension(:), allocatable :: compressedTensor
  integer i,j,dirs,k
  real(kind=RAdKind), external :: explicitSinProdTensorEntry
  real(kind=RAdKind) :: entry
  n=3
  d=3
  allocate(x(n))
  allocate(xv(n))
  ! argument values
  do i=1,n
     xv(i)=1.0+.1*i
     x(i)=xv(i)
  end do
  ! initialize the tensor context
  call setNumberOfIndependents(T,n)
  call setHighestDerivativeDegree(T,d)
  dirs=getDirectionCount(T)
  allocate(SeedMatrix(n,dirs))
  call getSeedMatrix(T,SeedMatrix)
  do i=1,dirs
     do j=1,n
        call RAset(x(j),i,1,REAL(SeedMatrix(j,i),kind=RAdKind))
     end do
  end do
  ! compute the target function
  call head(x,y)
  ! transfer the taylor coefficients
  allocate(TaylorCoefficients(d,dirs))
  do i=1,d
     do j=1,dirs
        call RAget(y,j,i,TaylorCoefficients(i,j))
     end do
  end do
  call setTaylorCoefficients(T,TaylorCoefficients)
  do k=1, d
     write(*,'(A,I2)') "order: ", k
     call reset(Thelper)
     call setNumberOfIndependents(Thelper,n)
     call setHighestDerivativeDegree(Thelper,k)
     dirs=getDirectionCount(Thelper)
     if (allocated(compressedTensor)) deallocate(compressedTensor)
     allocate(compressedTensor(dirs))
     if (allocated(SeedMatrixHelper)) deallocate(SeedMatrixHelper)
     allocate(SeedMatrixHelper(n,dirs))
     call getSeedMatrix(Thelper,SeedMatrixHelper)
     ! compute the compressedTensor
     call getCompressedTensor(T,k,compressedTensor)
     do i=1,dirs
        entry=explicitSinProdTensorEntry(n,xv,SeedMatrixHelper(:,i))
        write(*,'(A)',ADVANCE='NO') 'T'
        do j=1,n
           write(*,'(A,I2,A)',ADVANCE='NO') '[',SeedMatrixHelper(j,i),']'
        end do
        write(*,'(A,E25.17E3)',ADVANCE='NO') ' ',compressedTensor(i)
        write(*,'(A,E25.17E3)',ADVANCE='NO') ' ?= ', entry
        if (abs(compressedTensor(i)-entry)>1.0D-6) then 
           write(*,'(A,E25.17E3)') ' diff is: ', abs(compressedTensor(i)-entry)
           stop 1
        else
           write(*,'(A)') ' ok'
        end if
     end do
  end do
  deallocate(compressedTensor)
  deallocate(TaylorCoefficients)
  deallocate(SeedMatrix)

end program driver


